sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -2943,3 +2943,87 @@ case class Sentences(
     copy(str = newFirst, language = newSecond, country = newThird)
 
 }
+
+/**
+ * Splits a given string by a specified delimiter.
+ */
+case class SplitByDelimiter(
+    str: Expression,
+    delimiter: Expression)
+  extends BinaryExpression with NullIntolerant {
+  override def dataType: DataType = ArrayType(StringType, containsNull = false)
+  override def left: Expression = str
+  override def right: Expression = delimiter
+
+  override def nullSafeEval(string: Any, delimiter: Any): Any = {
+    val strings = {
+      // if delimiter is empty string, skip the regex based splitting directly as regex
+      // treats empty string as matching anything, thus use the input directly.
+      if (delimiter.asInstanceOf[UTF8String].numBytes() == 0) {
+        Array(string)
+      } else {
+        string.asInstanceOf[UTF8String].splitSQL(
+          delimiter.asInstanceOf[UTF8String], -1)
+      }
+    }
+    new GenericArrayData(strings.asInstanceOf[Array[Any]])
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val arrayClass = classOf[GenericArrayData].getName
+    nullSafeCodeGen(ctx, ev, (str, delimiter) => {
+      if (delimiter.asInstanceOf[UTF8String].numBytes() == 0) {
```

Comment:
I'm surprised the test passed. `delimiter` here is String type, which indicates the variable name of delimiter in the generated java code, please check `nullSafeCodeGen`

Reply:
Actually, can we put this empty delimiter handling in `UTF8String.splitSQL`? then we can make this expression extends `RuntimeReplaceable` with `Invoke`, and we don't need to implement the complicated codegen part.

Reply:
That requires to make the `splitSQL` as a static function. 

I tried to use this code:

```
    StaticInvoke(
      classOf[UTF8String],
      ArrayType.apply(StringType),
      "splitSQL",
      Seq(str, delimiter, Literal.create(-1, IntegerType)),
      Seq(StringType, StringType, IntegerType))
```
This gives me the following error message:

```
class [Lorg.apache.spark.unsafe.types.UTF8String; cannot be cast to class org.apache.spark.sql.catalyst.util.ArrayData ([Lorg.apache.spark.unsafe.types.UTF8String; and org.apache.spark.sql.catalyst.util.ArrayData are in unnamed module of loader 'app')
java.lang.ClassCastException: class [Lorg.apache.spark.unsafe.types.UTF8String; cannot be cast to class org.apache.spark.sql.catalyst.util.ArrayData ([Lorg.apache.spark.unsafe.types.UTF8String; and org.apache.spark.sql.catalyst.util.ArrayData are in unnamed module of loader 'app')
	at org.apache.spark.sql.catalyst.expressions.ElementAt.$anonfun$doElementAt$1(collectionOperations.scala:2178)
	at org.apache.spark.sql.catalyst.expressions.ElementAt.nullSafeEval(collectionOperations.scala:2173)
	at org.apache.spark.sql.catalyst.expressions.BinaryExpression.eval(Expression.scala:601)
```

Reply:
With your suggestion of moving the empty check into `splitSQL`, the codegen part and `nullSafeEval` part become much simpler already. What do you think?

Reply:
There is a `Invoke` expression for calling non-static functions.

Reply:
```
    Invoke(
      str,
      "splitSQL",
      dataType,
      Seq(delimiter, Literal.create(-1, IntegerType)),
      Seq(StringType, IntegerType))
  }
```

I see. It is easy to use `Invoke`.

However, the next challenge is the original expected output of the `eval` is a scala class `GenericArrayData`. The original `eval` just use `new GenericArrayData(strings.asInstanceOf[Array[Any]])` to create that.

Do you know how to make this work with `Invoke`? I cannot add `new GenericArrayData` to `splitSQL` because `splitSQL` is in a Java module which does not depend on scala.

Reply:
I see, then let's forget about `Invoke`. The codegen implementation should be simpler now.